snapcraft/internal/meta.py

```diff
@@ -301,50 +301,68 @@ def _wrap_app(self, name, app):
                 raise EnvironmentError(
                     'The specified command {!r} defined in the app {!r} '
                     'does not exist or is not executable'.format(str(e), name))
-        desktop_file = app.pop('desktop', '')
-        if desktop_file:
-            self._reformat_desktop(name, desktop_file)
-
-    def _reformat_desktop(self, name, desktop_file):
-        gui_dir = os.path.join(self.meta_dir, 'gui')
-        desktop_file_path = os.path.join(self._snap_dir, desktop_file)
-        if not os.path.exists(desktop_file_path):
+        desktop_file_name = app.pop('desktop', '')
+        if desktop_file_name:
+            desktop_file = _DesktopFile(
+                name=name, filename=desktop_file_name,
+                snap_name=self._config_data['name'], prime_dir=self._snap_dir,
+                gui_dir=os.path.join(self.meta_dir, 'gui'))
+            desktop_file.parse_and_reformat()
+            desktop_file.write()
+
+
+class _DesktopFile:
+
+    def __init__(self, *, name, filename, snap_name, prime_dir, gui_dir):
```

Comment:
Agreed for gui_dir, not so much for filename as it’s being used in all three functions (__init__, parse_and_reformat, write), so I see little point in passing it as a parameter to all of them. Plus, I think it makes more sense semantically speaking to have filename as a parameter of __init__, given that a _DesktopFile instance operates on one given desktop file.